File: Flix/app/src/main/java/fin/phoenix/flix/api/ProfileService.kt
package fin.phoenix.flix.api

import com.google.gson.annotations.SerializedName
import fin.phoenix.flix.data.Seller
import fin.phoenix.flix.data.User
import fin.phoenix.flix.data.UserAbstract
import fin.phoenix.flix.util.Resource
import retrofit2.Response
import retrofit2.http.*

/**
 * 用户相关API服务接口
 */
interface ProfileService {
    /**
     * 获取用户完整资料
     */
    @GET("profile/user/{userId}")
    suspend fun getUserProfile(@Path("userId") userId: String): Response<GenericApiResponse<User>>

    /**
     * 获取商家信息
     */
    @GET("profile/seller/{userId}")
    suspend fun getSellerProfile(@Path("userId") userId: String): Response<GenericApiResponse<Seller>>

    /**
     * 获取用户简要信息
     */
    @GET("profile/abstract/{userId}")
    suspend fun getUserAbstract(@Path("userId") userId: String): Response<GenericApiResponse<UserAbstract>>
    
    /**
     * 获取热门卖家列表
     */
    @GET("profile/popular")
    suspend fun getPopularSellers(@Query("limit") limit: Int): Response<GenericApiResponse<List<UserAbstract>>>
    
    /**
     * 更新用户资料
     */
    @PUT("profile/update")
    suspend fun updateUserProfile(@Body user: ProfileUpdateRequest): Response<GenericApiResponse<User>>
    
    /**
     * 充值余额
     */
    @POST("profile/recharge")
    suspend fun rechargeBalance(
        @Query("userId") userId: String,
        @Query("amount") amount: Int
    ): Response<GenericApiResponse<User>>
    
    /**
     * 更新用户头像
     */
    @POST("profile/avatar")
    suspend fun updateAvatar(
        @Query("userId") userId: String,
        @Query("avatarUrl") avatarUrl: String
    ): Response<GenericApiResponse<User>>
    
    /**
     * 获取用户发布的商品
     */
    @GET("profile/{userId}/products")
    suspend fun getUserProducts(@Path("userId") userId: String): Response<GenericApiResponse<List<String>>>
    
    /**
     * 获取用户已售商品
     */
    @GET("profile/{userId}/sold")
    suspend fun getUserSoldProducts(@Path("userId") userId: String): Response<GenericApiResponse<List<String>>>
    
    /**
     * 获取用户购买的商品
     */
    @GET("profile/{userId}/purchased")
    suspend fun getUserPurchasedProducts(@Path("userId") userId: String): Response<GenericApiResponse<List<String>>>
    
    /**
     * 获取用户收藏的商品
     */
    @GET("profile/{userId}/favorites")
    suspend fun getUserFavorites(@Path("userId") userId: String): Response<GenericApiResponse<List<String>>>
}


data class ProfileUpdateRequest(
    @SerializedName("uid") val uid: String, // User ID, unique identifier (primary key)
    @SerializedName("phone_number") val phoneNumber: String, // User's phone number
    @SerializedName("user_name") val userName: String, // Username
    @SerializedName("avatar_url") val avatarUrl: String? = null, // User avatar image URL
    @SerializedName("addresses") val addresses: List<String> = emptyList(), // List of user addresses
    @SerializedName("current_address") val currentAddress: String? = null, // Current selected address
    @SerializedName("school_id") val schoolId: String? = null, // User's school ID
    @SerializedName("campus_id") val campusId: String? = null  // User's campus ID
)

fun User.toUpdateRequest(): ProfileUpdateRequest {
    return ProfileUpdateRequest(
        uid = uid,
        phoneNumber = phoneNumber,
        userName = userName,
        avatarUrl = avatarUrl,
        addresses = addresses,
        currentAddress = currentAddress,
        schoolId = schoolId,
        campusId = campusId
    )
}

fun Resource<User>.toProfileUpdateRequestResource(): Resource<ProfileUpdateRequest> {
    return when (this) {
        is Resource.Success -> Resource.Success(data.toUpdateRequest())
        is Resource.Error -> Resource.Error(message)
        is Resource.Loading -> Resource.Loading
    }
}

File: flix_backend/lib/flix_backend/profile_service.ex
defmodule FlixBackend.ProfileService do
  import Ecto.Query, warn: false
  alias FlixBackend.Repo
  alias FlixBackend.Data.User

  @doc """
  获取用户完整资料
  """
  def get_user_profile(user_id) do
    case Repo.get(User, user_id) do
      nil -> {:error, "用户不存在"}
      user -> {:ok, user}
    end
  end

  @doc """
  获取商家资料（去除敏感信息）
  """
  def get_seller_profile(user_id) do
    case Repo.get(User, user_id) do
      nil -> {:error, "用户不存在"}
      user ->
        seller = %{
          uid: user.uid,
          user_name: user.user_name,
          avatar_url: user.avatar_url,
          school_id: user.school_id,
          published_product_ids: user.published_product_ids,
          current_address: user.current_address,
          phone_number: user.phone_number,
          sold_product_ids: user.sold_product_ids,
          campus_id: user.campus_id,
          sold_count: length(user.sold_product_ids || [])
        }

        {:ok, seller}
    end
  end

  @doc """
  获取用户简要信息（头像、昵称、ID等）
  """
  def get_user_abstract(user_id) do
    case Repo.get(User, user_id) do
      nil -> {:error, "用户不存在"}
      user ->
        user_abstract = %{
          uid: user.uid,
          user_name: user.user_name,
          avatar_url: user.avatar_url
        }

        {:ok, user_abstract}
    end
  end

  @doc """
  获取热门卖家
  """
  def get_popular_sellers(limit) do
    query = from u in User,
            order_by: [desc: fragment("array_length(?, 1)", u.sold_product_ids)],
            limit: ^limit

    sellers = Repo.all(query)
    |> Enum.map(fn user ->
      %{
        uid: user.uid,
        user_name: user.user_name,
        avatar_url: user.avatar_url,
        sold_count: length(user.sold_product_ids || [])
      }
    end)

    {:ok, sellers}
  end

  @doc """
  更新用户资料
  """
  def update_user_profile(user_params) do
    user_id = user_params["userId"] || user_params["user_id"]

    with %User{} = user <- Repo.get(User, user_id) do
      # 移除userId/user_id,确保不会更新ID
      attrs = case user_params do
        %{"userId" => _} -> Map.delete(user_params, "userId")
        %{"user_id" => _} -> Map.delete(user_params, "user_id")
        _ -> user_params
      end

      # 只允许更新特定字段
      allowed_attrs = Map.take(attrs, [
        "user_name", "avatar_url", "school_id", "campus_id"
      ])

      user
      |> User.changeset(allowed_attrs)
      |> Repo.update()
    else
      nil -> {:error, "用户不存在"}
    end
  end

  @doc """
  充值余额
  """
  def recharge_balance(user_id, amount) when is_integer(amount) and amount > 0 do
    with %User{} = user <- Repo.get(User, user_id) do
      user
      |> User.changeset(%{balance: user.balance + amount})
      |> Repo.update()
    else
      nil -> {:error, "用户不存在"}
    end
  end

  def recharge_balance(_user_id, _amount), do: {:error, "无效的充值金额"}

  @doc """
  更新用户头像
  """
  def update_avatar(user_id, avatar_url) when is_binary(avatar_url) do
    with %User{} = user <- Repo.get(User, user_id) do
      user
      |> User.changeset(%{avatar_url: avatar_url})
      |> Repo.update()
    else
      nil -> {:error, "用户不存在"}
    end
  end

  def update_avatar(_user_id, _avatar_url), do: {:error, "无效的头像URL"}

  @doc """
  获取用户发布的商品ID列表
  """
  def get_user_products(user_id) do
    with %User{} = user <- Repo.get(User, user_id) do
      {:ok, user.published_product_ids || []}
    else
      nil -> {:error, "用户不存在"}
    end
  end

  @doc """
  获取用户已售商品ID列表
  """
  def get_user_sold_products(user_id) do
    with %User{} = user <- Repo.get(User, user_id) do
      {:ok, user.sold_product_ids || []}
    else
      nil -> {:error, "用户不存在"}
    end
  end

  @doc """
  获取用户购买的商品ID列表
  """
  def get_user_purchased_products(user_id) do
    with %User{} = user <- Repo.get(User, user_id) do
      {:ok, user.purchased_product_ids || []}
    else
      nil -> {:error, "用户不存在"}
    end
  end

  @doc """
  获取用户收藏的商品ID列表
  """
  def get_user_favorites(user_id) do
    with %User{} = user <- Repo.get(User, user_id) do
      {:ok, user.favorite_product_ids || []}
    else
      nil -> {:error, "用户不存在"}
    end
  end

  @doc """
  获取用户列表（管理员功能）
  """
  def list_users(offset, limit) do
    query = from u in User

    # 计算总数
    total_count = Repo.aggregate(query, :count, :uid)

    # 添加分页和排序
    query = from u in query,
            order_by: [desc: u.inserted_at],
            limit: ^limit,
            offset: ^((offset - 1) * limit)

    users = Repo.all(query)

    {:ok, users, total_count}
  end
end
